nx/lib/nx/defn/expr.ex

```diff
@@ -984,9 +991,16 @@ defmodule Nx.Defn.Expr do
 
   defp constant(%{type: type, shape: shape} = out, number) do
     number =
-      if is_integer(number) and Nx.Type.float?(type),
-        do: Complex.multiply(1.0, number),
-        else: number
+      cond do
+        is_integer(number) and Nx.Type.float?(type) ->
+          Complex.multiply(1.0, number)
+
+        not is_integer(number) and Nx.Type.integer?(type) ->
+          raise "value #{inspect(number)} is not valid for constant of type #{inspect(type)}"
```

Comment:
Let me know if having this assertion is worth it.

Reply:
```suggestion
        not is_integer(number) and Nx.Type.integer?(type) ->
          raise ArgumentError, "value #{inspect(number)} is not valid for constant of type #{inspect(type)}"
```